File: Facade_Pattern/go/Facade_Pattern.go
package Facade_Pattern

import "fmt"

type HardDrive struct{}

func (*HardDrive) load() {
	fmt.Println("hard drive loading")
}

type Memory struct{}

func (*Memory) read() {
	fmt.Println("memory reading")
}

type CPU struct{}

func (*CPU) execute() {
	fmt.Println("cpu executing")
}

type Computer struct {
	hardDrive HardDrive
	memory    Memory
	cpu       CPU
}

func (this *Computer) start() {
	this.hardDrive.load()
	this.cpu.execute()
	this.memory.read()
}

File: Facade_Pattern/cpp/Facade_Pattern.cc
#include <iostream>
using namespace std;

class HardDrive {
public:
    void load() {
        cout << "hard drive loading" << endl;
    }
};

class Memory {
public:
    void read() {
        cout << "memory reading" << endl;
    }
};

class CPU {
public:
    void execute() {
        cout << "cpu executing" << endl;
    }
};

class Computer {
public:
    void start() {
        hardDrive.load();
        cpu.execute();
        memory.read();
    }
public:
    HardDrive hardDrive;
    Memory memory;
    CPU cpu;
};

void test_Facade() {
    Computer c;
    c.start();
}
int main() {
    test_Facade();
    return 0;
}